host/src/attribute_server.rs

```diff
@@ -1,31 +1,202 @@
+                    //  trace!("[server] empty slot: connecting");
+                    client.is_connected = true;
+                    client.address = *peer_address;
+                    return;
+                }
+            }
+            trace!("[server] all slots full...");
+            // if we got here all slots are full; replace the first disconnected client
+            for (client, table) in n.iter_mut() {
+                if !client.is_connected {
+                    trace!("[server] booting disconnected peer {:?}", client.address);
+                    client.is_connected = true;
+                    client.address = *peer_address;
+                    // erase the previous client's config
+                    table.disable_all();
+                    return;
+                }
+            }
+            // Should be unreachable if the max connections (CN) matches that defined
+            // in HostResources...
+            warn!("[server] unable to obtain CCCD slot");
+        })
+    }
+
+    fn disconnect(&self, peer_address: &BdAddr) {
+        self.state.lock(|n| {
+            let mut n = n.borrow_mut();
+            for (client, _) in n.iter_mut() {
+                if client.address == *peer_address {
+                    client.is_connected = false;
+                    break;
+                }
+            }
+        })
+    }
+
+    fn get_value(&self, peer_address: &BdAddr, cccd_handle: u16) -> Option<[u8; 2]> {
+        self.state.lock(|n| {
+            let n = n.borrow();
+            for (client, table) in n.iter() {
+                if client.address == *peer_address {
+                    return table.get_raw(cccd_handle);
+                }
+            }
+            None
+        })
+    }
+
+    fn set_notify(&self, peer_address: &BdAddr, cccd_handle: u16, is_enabled: bool) {
+        self.state.lock(|n| {
+            let mut n = n.borrow_mut();
+            for (client, table) in n.iter_mut() {
+                if client.address == *peer_address {
+                    table.set_notify(cccd_handle, is_enabled);
+                    break;
+                }
+            }
+        })
+    }
+
+    fn should_notify(&self, peer_address: &BdAddr, cccd_handle: u16) -> bool {
+        self.state.lock(|n| {
+            let n = n.borrow();
+            for (client, table) in n.iter() {
+                if client.address == *peer_address {
+                    return table.should_notify(cccd_handle);
+                }
+            }
+            false
+        })
+    }
 }
 
 /// A GATT server capable of processing the GATT protocol using the provided table of attributes.
-pub struct AttributeServer<'values, M: RawMutex, const MAX: usize> {
-    pub(crate) table: AttributeTable<'values, M, MAX>,
-    pub(crate) notification: Mutex<M, RefCell<NotificationTable<MAX_NOTIFICATIONS>>>,
+pub struct AttributeServer<'values, M: RawMutex, const AT: usize, const CT: usize, const CN: usize> {
```

Comment:
I'd rather see some fully spelled names for these, or at the minimum some docs explaining what they mean.